File: backend/models/property.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type Property struct {
	ID            primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	Title         string             `json:"title" bson:"title" binding:"required"`
	Description   string             `json:"description" bson:"description"`
	Type          string             `json:"type" bson:"type"` // "shop", "partition", "office", "warehouse"
	Location      Location           `json:"location" bson:"location"`
	RentPerMonth  float64            `json:"rentPerMonth" bson:"rentPerMonth" binding:"required"`
	SecurityDeposit float64          `json:"securityDeposit" bson:"securityDeposit"`
	Area          float64            `json:"area" bson:"area"` // in sq ft
	Amenities     []string           `json:"amenities" bson:"amenities"`
	Images        []string           `json:"images" bson:"images"`
	Status        string             `json:"status" bson:"status"` // "available", "occupied", "maintenance"
	Features      PropertyFeatures   `json:"features" bson:"features"`
	CreatedBy     primitive.ObjectID `json:"createdBy" bson:"createdBy"`
	CreatedAt     time.Time          `json:"createdAt" bson:"createdAt"`
	UpdatedAt     time.Time          `json:"updatedAt" bson:"updatedAt"`
}

type Location struct {
	Street     string  `json:"street" bson:"street"`
	Area       string  `json:"area" bson:"area"`
	City       string  `json:"city" bson:"city"`
	District   string  `json:"district" bson:"district"`
	State      string  `json:"state" bson:"state"`
	Pincode    string  `json:"pincode" bson:"pincode"`
	Landmark   string  `json:"landmark" bson:"landmark"`
	Latitude   float64 `json:"latitude" bson:"latitude"`
	Longitude  float64 `json:"longitude" bson:"longitude"`
}

type PropertyFeatures struct {
	HasElectricity    bool `json:"hasElectricity" bson:"hasElectricity"`
	HasWater          bool `json:"hasWater" bson:"hasWater"`
	HasParking        bool `json:"hasParking" bson:"hasParking"`
	HasSecurity       bool `json:"hasSecurity" bson:"hasSecurity"`
	HasAC             bool `json:"hasAC" bson:"hasAC"`
	HasWiFi           bool `json:"hasWiFi" bson:"hasWiFi"`
	IsFurnished       bool `json:"isFurnished" bson:"isFurnished"`
	HasRestroom       bool `json:"hasRestroom" bson:"hasRestroom"`
}

File: backend/models/Property.js
const mongoose = require('mongoose');

const propertySchema = new mongoose.Schema({
  title: {
    type: String,
    required: [true, 'Property title is required'],
    trim: true,
    maxlength: [100, 'Title cannot exceed 100 characters']
  },
  description: {
    type: String,
    required: [true, 'Property description is required'],
    maxlength: [1000, 'Description cannot exceed 1000 characters']
  },
  type: {
    type: String,
    required: [true, 'Property type is required'],
    enum: ['apartment', 'villa', 'room', 'pg', 'hostel', 'house']
  },
  location: {
    address: {
      type: String,
      required: [true, 'Address is required']
    },
    city: {
      type: String,
      required: [true, 'City is required']
    },
    state: {
      type: String,
      default: 'Rajasthan'
    },
    pincode: {
      type: String,
      required: [true, 'Pincode is required'],
      match: [/^\d{6}$/, 'Please provide a valid pincode']
    },
    coordinates: {
      latitude: Number,
      longitude: Number
    }
  },
  rent: {
    amount: {
      type: Number,
      required: [true, 'Rent amount is required'],
      min: [500, 'Rent must be at least ₹500']
    },
    currency: {
      type: String,
      default: 'INR'
    },
    deposit: {
      type: Number,
      required: [true, 'Security deposit is required']
    },
    maintenance: {
      type: Number,
      default: 0
    }
  },
  amenities: [{
    type: String,
    enum: [
      'wifi', 'parking', 'ac', 'furnished', 'gym', 'swimming_pool',
      'security', 'elevator', 'balcony', 'garden', 'power_backup',
      'water_supply', 'cooking_allowed', 'pets_allowed'
    ]
  }],
  specifications: {
    bedrooms: {
      type: Number,
      required: true,
      min: [0, 'Bedrooms cannot be negative']
    },
    bathrooms: {
      type: Number,
      required: true,
      min: [1, 'At least 1 bathroom is required']
    },
    area: {
      value: {
        type: Number,
        required: true
      },
      unit: {
        type: String,
        enum: ['sqft', 'sqm'],
        default: 'sqft'
      }
    },
    floor: Number,
    totalFloors: Number,
    furnished: {
      type: String,
      enum: ['fully', 'semi', 'unfurnished'],
      default: 'unfurnished'
    }
  },
  images: [{
    url: String,
    caption: String,
    isPrimary: { type: Boolean, default: false }
  }],
  availability: {
    isAvailable: {
      type: Boolean,
      default: true
    },
    availableFrom: {
      type: Date,
      default: Date.now
    },
    preferredTenant: {
      type: String,
      enum: ['any', 'family', 'bachelor', 'student', 'working_professional'],
      default: 'any'
    }
  },
  owner: {
    name: String,
    phone: String,
    email: String,
    isVerified: { type: Boolean, default: false }
  },
  createdBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  isActive: {
    type: Boolean,
    default: true
  },
  views: {
    type: Number,
    default: 0
  },
  featured: {
    type: Boolean,
    default: false
  }
}, {
  timestamps: true
});

// Index for location-based searches
propertySchema.index({ 'location.city': 1, 'location.state': 1 });
propertySchema.index({ 'rent.amount': 1 });
propertySchema.index({ type: 1 });
propertySchema.index({ 'availability.isAvailable': 1 });

module.exports = mongoose.model('Property', propertySchema);